### PR TITLE
UseNewEnvironment restores environment variables for child process

### LINKED ISSUE #3545: Start-Process fails to launch new powershell.exe instance with -UseNewEnvironment switch
Steps to reproduce
------------------
I dont know why would anyone do this, but here it goes: you cannot launch another `powershell.exe` using `Start-Process` cmdlet with `-UseNewEnvironment` switch.

Windows PowerShell:
```
Start-Process -FilePath C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -NoNewWindow -UseNewEnvironment
```

PowerShell Core (Debug build):
```
Start-Process -FilePath .\powershell.exe -NoNewWindow -UseNewEnvironment
```

Expected behavior
-----------------
A new instance of powershell.exe console should start.

Actual behavior
---------------
Windows PowerShell outputs error and dies:
```
Internal Windows PowerShell error.  Loading managed Windows PowerShell failed with error 8009001d.
```

PowerShell Core (Debug build) outputs more errors + assert message and dies:
```
Error resolving full path []
Error resolving full path []
PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

The shell cannot be started. A failure occurred during initialization:
ASSERT: caller makes sure it's not null or empty
```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      5.1.14409.1005
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14409.1005
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.0.0-alpha
GitCommitId                    v6.0.0-alpha.18-9-g8d4db01a5d4b5dcd981ef216d3d7f24b484c7c60-dirty
WSManStackVersion              3.0
CLRVersion
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   3.0.0.0
PSRemotingProtocolVersion      2.3
PSEdition                      Core
SerializationVersion           1.1.0.1
```

### LINKED ISSUE #4671: Start-Process -UseNewEnvironment provides an environment that is missing crucial standard environment variables while not supporting passing a new environment
On both Windows and Unix platforms, `Start-Process -UseNewEnvironment` results in an environment that is missing crucial standard environment variables, making the new environment virtually useless, while not providing a mechanism to define a new environment:

* On Windows, `-UseNewEnvironment` defines _only_ the variables that are _explicitly_ defined, as displayed in System Properties (`sysdm.cpl`), with crucial, usually automatically defined variables such as `$env:ProgramFiles` missing, and `$env:USERNAME` unexpectedly containing `SYSTEM`.

     * This even prevents another `powershell` instance from being started this way - see https://github.com/PowerShell/PowerShell/issues/3545#issuecomment-320699488 and below.

* On Unix, an _empty_ environment is essentially passed (with _no_ environment variables defined at all, except `$env:PSModulePath`).

    * This even prevents invoking any executable by filename only, given that `$env:PATH` is undefined - see below.

    * While this behavior is similar to POSIX [`env -i`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/env.html) on Unix platforms, the `env` utility notably includes a mechanism to provide a _new_ environment via `name=value` pairs, which `Start-Process` lacks.

---

Possible solutions:

* Repurpose `-UseNewEnvironment` to not start with a blank slate / crucial variables missing, but to provide the same environment that a shell would see when directly launched by the system (in other words: any environment-variable modifications made by the calling shell would be ignored).

* Additionally, provide a way to pass a custom environment, applied _on top_ of the current or pristine (`-UseNewEnvironment`) environment:

   * E.g., a new `-Environment <hashtable>`  / `-Environment <Collections.DictionaryEntry[]>` parameter could be used.

    * If someone truly wanted an _empty_ environment, they could start with `$emptyEnv = (Get-ChildItem env:); $emptyEnv | % { $_.Value = $null }` and pass `-Environment $emptyEnv`.

---

The `-Environment` feature would allow for an - incomplete - approximation of the convenient ad-hoc, command-scoped environment-variable definition feature that POSIX-like shells such as `bash` offer, where you can prepend one or more `name=value` pairs to a command:

```sh
# Bash (any POSIX-compatible shell)
# Defines $env:FOO as 'bar', but *only for the some-utility child process*.
FOO=bar some-utility 666
```

The PS approximation would be:

```powershell
Start-Process -Wait  -Environment @{ FOO = 'bar' } some-utility -Args 666
```

That said, a crucial limitation is that use of `Start-Process` makes the external utility operate outside PowerShell's streams, so the only way to provide input / collect output is via the `-Redirect*` parameters.

Current behavior
------------------

```powershell
Start-Process -UseNewEnvironment -Wait -NoNewWindow powershell -args '-Command', 'gci env:; whoami'
```

* On Windows: PowerShell refuses to start - see #3545; note that adding `-LoadUserProfile` makes no difference.

```none
Internal Windows PowerShell error.  Loading managed Windows PowerShell failed with error 8009001d.
```

* On Unix: PowerShell starts, lists `$env:PSModulePath` as the _only_ environment variable - with a seemingly temporary user account's module directory prepended -  and the `whoami` invocation fails, because it cannot be located in the absence of a suitable `$env:PATH`.

```none

Name                           Value                                                                                                                           
----                           -----                                                                                                                           
PSModulePath                   /tmp/ba38e79f-40c2-440e-ae08-7cf32e0708e1/.local/share/powershell/Modules:/usr/local/share/powershell/Modules:/opt/microsoft/...
whoami : The term 'whoami' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path 
was included, verify that the path is correct and try again.
At line:1 char:21
+ Get-ChildItem env:; whoami
+                     ~~~~~~
    + CategoryInfo          : ObjectNotFound: (whoami:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.5 on macOS 10.12.6
PowerShell Core v6.0.0-beta.5 on Ubuntu 16.04.3 LTS
PowerShell Core v6.0.0-beta.5 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
Windows PowerShell v5.1.15063.483 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
```

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

- Save environment variables at startup
- Restore these saved environment variables for child process is UseNewEnvironment parameter is used in Start-Process cmdlet

Formally it is a breaking change in grey area.

I hope we do not load any additional dlls with the change at startup and slowdown the startup scenario will be minimal.

## PR Context

Fix #4671
Fix #3545

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 1]
```diff
@@ -65,6 +65,18 @@ internal static void Init()
                 LanguagePrimitives.GetEnumerator(null);
             });
         }
+
+        internal static IDictionary InitialEnvironmentVariables
+        {
+            get => _environmentVariables;
+        }
+
+        private static IDictionary _environmentVariables;
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: Style: usually a field definition appears before it is used.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs [hunk 2]
```diff
@@ -1908,9 +1908,15 @@ protected override void BeginProcessing()
 
                 if (_UseNewEnvironment)
                 {
-                    startInfo.EnvironmentVariables.Clear();
-                    LoadEnvironmentVariable(startInfo, Environment.GetEnvironmentVariables(EnvironmentVariableTarget.Machine));
-                    LoadEnvironmentVariable(startInfo, Environment.GetEnvironmentVariables(EnvironmentVariableTarget.User));
+                    // Use environment variables from saved at startup dictionary
+                    var envVariables = startInfo.EnvironmentVariables;
+                    envVariables.Clear();
+                    IDictionaryEnumerator e = System.Management.Automation.Runspaces.EarlyStartup.InitialEnvironmentVariables.GetEnumerator();
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs]: Just curious, why did you use an enumerator object explicitly rather than:
```
foreach (DictionaryEntry entry in EarlyStartup.InitialEnvironmentVariables)
{
    ...
}
```
?

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/InitialSessionState.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/InitialSessionState.cs]: Fixed.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs (hunk 2 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs]: I copied the code from CoreFX.

Fixed.

**Review by PaulHigin — APPROVED** [review]
LGTM.  But committee will have to decide whether to take this breaking change.

**Comment by joeyaiello** [comment]
Will talk about this with @PowerShell/powershell-committee today, but I think this is a worthwhile change given it's what's the norm across Unix systems. New processes should be expected to set environment state they don't want to be handed by parent processes. 

Would love to understand how this works on Windows today, though, as it was my expectation that things already worked this way:

```powershell
PowerShell 6.2.1
Copyright (c) Microsoft Corporation. All rights reserved.

https://aka.ms/pscore6-docs
Type 'help' to get help.

Loading personal and system profiles took 883ms.
C:\Users\jaiello> $env:foo = "bar"
C:\Users\jaiello> powershell
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

Try the new cross-platform PowerShell https://aka.ms/pscore6

PS C:\Users\jaiello> $env:foo
bar
```

This also is true when I use `Start-Process powershell` and get a new terminal.

**Comment by joeyaiello** [comment]
I misunderstood the original intent of this PR, apologies (reading too fast).

@PowerShell/powershell-committee discussion on this was interesting.

First, it looks like `Start-Process -UseNewEnvironment` on Windows gives the equivalent of a fresh login (i.e. Machine and User variables as they're stored in the registry). 

Additionally, this caching behavior can be somewhat non-deterministic when I'm executed from multiple shells. For instance, I could:

1. start my login shell as Bash
1. modify my environment
1. start pwsh
1. modify my environment
1. start pwsh
1. modify my environment
1. `Start-Process -UseNewEnvironment pwsh`

I'm now getting the modified state post-step 4 instead of the original login shell environment post-step 1 (as I'd expect to get on Windows today). 

Therefore, we think the behavior that folks would expect from this is to get the login shell behavior with `-UseNewEnvironment`. On Linux, this would match @rjmholt 's behavior for `pwsh -l` where we use `sh -l` on Linux and `zsh -l` on macOS each in conjunction with `-c export` to retrieve explicitly and only the exported environment variables from the login shell invocation. 

If you disagree, can you help us understand a scenario where you'd actually want the exact environment state from two shells up your stack?

**Comment by mklement0** [comment]
@joeyaiello:

The determinism I was going for was:

* For the child process about to be created, provide the same environment that the current process saw _when it started_.

* In other words: Ignore all _in-session_ modifications made by the current process.

That is indeed not the same as "provide the same environment a new shell process would see if created directly", not via another shell instance, but, _pragmatically  speaking_:

* It may be _good enough_ (you don't typically _nest_ shells, and the need to ignore _in-session_ changes strikes me as the typical use case)

* It is easy to implement consistently, in a platform-neutral manner.

----

As for "provide the same environment a new shell process would see if created directly" functionality:

If it can be done, consistently across all platforms, I agree that it is the better choice.

* Windows (where the feature is [currently half broken](https://github.com/PowerShell/PowerShell/issues/4671#issue-252911712), also in Windows PowerShell):
 
  * @geelaw mentions in https://github.com/PowerShell/PowerShell/issues/3545#issuecomment-398105462 that the [`CreateEnvironmentBlock`](https://docs.microsoft.com/en-us/windows/win32/api/userenv/nf-userenv-createenvironmentblock) WinAPI function may be the right solution, but it also states:
  > User-specific environment variables such as `%USERPROFILE%` are set only when the user's profile is loaded. To load a user's profile, call the `LoadUserProfile` function.

  * @SteveL-MSFT just posted a thought at https://github.com/PowerShell/PowerShell/issues/4671#issuecomment-543346250:

> on Windows, the current code clears the env block, then fills it from the system profile (hence username is system) and then the user profile overwriting any that exist in the system. Hence Path is incomplete. It seems the fix here is to merge Path (append user PATH to end of machine PATH). USERNAME should be special cased to the user. 

* Unix-like platforms:

  * macOS: A login-shell like environment seems appropriate - but the CLI's `-l` workaround (launching via a hidden `sh` instance) is not really an option here; not sure how to get the right environment directly.

  * Unix: ? (No personal knowledge)

**Comment by mklement0** [comment]
P.S.: Forgot to mention that `cmd.exe`'s `start /I` feature - which is presumably what `Start-Process -UseNewEnvironment` was inspired by - indeed uses the use-the-shell's-own-startup-environment approach, as originally suggested:

```powershell
# Opens a new cmd.exe window that DOES see $env:FOO,
# because it was present in the startup environment in the intermediate cmd.exe instance,
# but not $env:FOO2
$env:FOO='bar'; cmd /c "set FOO2=no & start /I cmd /k echo %FOO%-%FOO2%"
```

**Comment by iSazonov** [comment]
Replaced by #10830

**Comment by mklement0** [comment]
Thinking about this some more:

* **Unix** platforms have _no_ native programmatic mechanism that I know of for "give me a pristine but functional environment":

  * Utility `env` with `-i` _radically clears_ the environment, which makes for a non-functional environment that `bash` itself happens to compensate for by having _hard-coded defaults_.

  * Someone who truly needs to start with a pristine environment can use `env -i` on Unix and manually reconstruct the essential system variables, if really needed - and using `env -i`  is also an option _from PowerShell_.

  * I'm not sure anyone would miss that ability on Windows, since it has never been there.

* On **Windows**, the original feature _similar_ to "give me a pristine but functional environment" was actually "give me a copy of _this process'_ original environment, without any _in-session_ changes" in the form of `cmd.exe`'s `start /I`

  * Windows PowerShell _tried to_ implement "give me a pristine but functional environment", but failed to do so (it is what #10830 now tries to fix).

---

In short: There's no _existing_ PowerShell functionality that we're really beholden to, on either platform.

Based on the pragmatic arguments made above:

* "give me a copy of _this process'_ original environment, before _in-session_ changes" is _typically_ the same as the pristine-environment feature.

* the former is easy to explain, easy to document, easy to implement across platforms

my vote is therefore to implement the "give me a copy of _this process'_ original environment, without any _in-session_ changes"  - unless someone knows how to - robustly, in a conceptually clean manner - implement the pristine-environment feature on macOS and Linux.

I think that is preferable to having the pristine-environment feature, but _on Windows only_.

**Comment by iSazonov** [comment]
@mklement0 Thanks for great comment!

As for Unix-s my current thought is to use Jim's workaround for login shell.

**Comment by mklement0** [comment]
@iSazonov:

A _login_ shell is only appropriate for macOS, not Linux.

In case you're thinking of the `/bin/sh -c 'exec pwsh ...`  workaround (please clarify):

That will (a) require you to start with a cleared environment, as with `env -i`, which (b) will _not_ give you a functional environment, even with `-l`:

```sh
$ env -i /bin/sh -l -c printenv
PATH=/usr/local/bin:/usr/bin:/bin:...
SHLVL=1
_=/usr/bin/printenv
```

As you can see, vital environment variables such as `USER`, `SHELL`, `LANG`, ... are missing.

**Comment by iSazonov** [comment]
> As you can see, vital environment variables such as USER, SHELL, LANG, ... are missing.

Where do these variables come from?

**Comment by mklement0** [comment]
It all relates to the unspoken premise of the "give me a pristine but functional environment" feature: A pristine environment _obtained how_?

In terms of _effect_, you'd want the same kind of environment you get:

* on Windows: when you open a new shell window via the Start Menu or the `Run` dialog, for instance.

* on Unix: when you use the default terminal program to open a new shell window.

My point is that we - certainly I - _don't know how to do that programmatically_ or even whether it's feasible on all Unix platforms (and at this point I'm also not convinced that the Windows-only #10830 is fully functional).

So, instead of going down that rabbit hole, we could implement the "give me a copy of this process' original environment, without any in-session changes" approach, document it, and be done.

I suspect no one will complain, because:

* it's the only _functioning_ approach currently available on Windows, via `cmd.exe`'s  `start /I`

* neither feature currently exists on Unix (only the clear-all feature does, which continues to be available via `env -i`)

* in the _typical_ use case it will be the same as "give me a pristine but functional environment"
